### PR TITLE
Editor: Use the same PostTemplatePanel between post and site editors

### PR DESCRIPTION
**Pull request by youknowriad**
Related #52632 
Builds on top of previous PR that align template modes between post and site editors.

## What?

This PR unifies the "template" panel/selectors on the document settings sidebar between the post and site editors. This means

 - You can now "swap templates" visually in the post editor.
 - You can now "reset to default template" in the post editor.
 - You can now enable "template preview" in the post editor.
 - You can now "create templates" in the site editor from the template panel
 - We maintain all of this code in a single place.

## Testing Instructions

Test the different flows related to the template panel in both post and site editors for classic and block themes.

### CODE REVIEW

File: packages/edit-post/src/components/sidebar/post-status/index.js [hunk 1]
```diff
@@ -62,14 +62,14 @@ export default function PostStatus() {
...
 						<PostURLPanel />
 						<PostSyncStatus />
-						<PostSticky />
-						<PostPendingStatus />
 						<PostFormat />
 						<PostSlug />
+						<PostTemplatePanel />
 						<PostAuthorPanel />
+						<PostSticky />
+						<PostPendingStatus />
```
Comment (youknowriad) [inline packages/edit-post/src/components/sidebar/post-status/index.js]: We have three kinds of panels right now:

 - Some are two column panels with a title and a dropdown on the right
 - Some are select taking the full with like "post author" and the classic theme's "page panel" 
 - Some are toggles like "sticky" and "pending status".
 
 I did a bit of shuffling here to make sure things looks correct (group similar panels together). I think as a follow-up we might want to transform panels like "post author" and "classic page panel" to two columns panels but this PR is very big already, so I just kept these as they are.

Comment (ellatrix) [reply packages/edit-post/src/components/sidebar/post-status/index.js]: But they are currently not grouped correctly in trunk?

Comment (youknowriad) [reply packages/edit-post/src/components/sidebar/post-status/index.js]: No, it's a bit random.

File: packages/edit-post/src/store/actions.js [hunk 2]
```diff
@@ -550,21 +549,14 @@ export const __unstableSwitchToTemplateMode =
...
-			.dispatch( coreStore )
-			.editEntityRecord( 'postType', post.type, post.id, {
-				template: savedTemplate.slug,
-			} );
-	};
+export function __unstableCreateTemplate() {
+	deprecated( "dispatch( 'core/edit-post' ).__unstableCreateTemplate", {
+		since: '6.5',
+	} );
+	return { type: 'NOTHING' };
```
Comment (ellatrix) [inline packages/edit-post/src/store/actions.js:543]: Lol

File: packages/editor/src/components/post-template/reset-default-template.js [hunk 3]
```diff
@@ -25,21 +25,19 @@ export default function ResetDefaultTemplate( { onClick } ) {
...
-						{ undoIgnore: true }
-					);
-					onClick();
-				} }
-			>
-				{ __( 'Use default template' ) }
-			</MenuItem>
-		</MenuGroup>
+		<MenuItem
+			onClick={ async () => {
```
Comment (ellatrix) [inline packages/editor/src/components/post-template/reset-default-template.js]: It was previously there, but this doesn't seem to need to be async.

File: packages/edit-post/src/components/sidebar/post-status/index.js (hunk 1 above)
Comment (ellatrix) [reply packages/edit-post/src/components/sidebar/post-status/index.js]: Just worried we are changing the control order for the main editor too, in a refactoring PR. But I don't mind.

**Review by ellatrix — APPROVED** [review]
Nice! Love the consistency between editors!

**Comment by jameskoster** [comment]
Nice, this is working well for me excerpt for one detail;

You can create a new template for pages where the template resolves automatically. E.g. if you edit the Posts page when Blog home exists, or if you edit the static homepage when Front page exists. The newly template even appears to get assigned to the page, but the frontend uses the correct template according to the hierarchy resulting in a mismatch.

Should we disable new template creation in those scenarios?

**Comment by youknowriad** [comment]
@jameskoster Good catch, I think it should be fixed now.

**Review by jameskoster — APPROVED** [review]
Works now 👍

**Comment by youknowriad** [comment]
I noticed thanks to the e2e test that this seems to break "custom templates" in classic themes.

**Comment by youknowriad** [comment]
Ok Thanks to e2e tests, I had to:

 - Restore the classic theme's template dropdown menu because it's a bit different, you need a select that mixes both php templates and block templates and you still want to be able to create block templates. (Hard to merge with the "swap template" flow) 
 - Had to move and unify some notifications about the template mode.

**Comment by youknowriad** [comment]
I still think we need to a bit more tests here. I'm almost certain that this breaks for contributors in the post editor but I don't have time today, I need to disconnect.

File: packages/edit-post/src/components/sidebar/post-status/index.js (hunk 1 above)
Comment (youknowriad) [reply packages/edit-post/src/components/sidebar/post-status/index.js]: I've restored the classic template selector, so I'm going to restore this as well.

**Comment by ramonjd** [comment]
Working as described for me 🎉 

> I'm almost certain that this breaks for contributors in the post editor

Do you mean the contributor role? If so, I tried creating pending posts as a contributor (who doesn't have access to template actions or the site editor), and then logged in with admin to swap templates and publish and couldn't see any errors.

**Review by andrewserong — COMMENTED** [review]
Very cool! So nice being able to toggle the template preview within the post editor 🎉

Also, updating the featured image via the Featured Image block in template preview mode updates the image in the Featured image panel and vise versa 👍

https://github.com/WordPress/gutenberg/assets/14988353/33f238ac-6a71-4068-81f7-e81e743020d2

I haven't run into any major issue so far in testing. One tiny thing I noticed is that in the template preview mode we still output the `padding-bottom: 40vh` rule, which means that the post appears to extend a bit further vertically than I expected. Is that intentional?

image

Another tiny thing is the padding / styling on this popover in Classic themes (looks like not quite enough padding on this PR?):

| This PR | Trunk |
| --- | --- |
| image | image |

**Comment by youknowriad** [comment]
Good catch @andrewserong both of these issues should be fixed.

**Comment by t-hamano** [comment]
I submitted #57112, which might be an issue related to this PR.

**Pull request merged**